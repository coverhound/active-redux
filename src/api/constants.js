export const API_CONFIGURE = 'AR_API_CONFIGURE';
export const API_CLEAR = 'AR_API_CLEAR';
export const API_HYDRATE = 'API_HYDRATE';

export const API_WILL_READ = 'AR_API_WILL_READ';
export const API_READ_DONE = 'AR_API_READ_DONE';
export const API_READ_FAILED = 'AR_API_READ_FAILED';

export const API_WILL_CREATE = 'AR_API_WILL_CREATE';
export const API_CREATE_DONE = 'AR_API_CREATE_DONE';
export const API_CREATE_FAILED = 'AR_API_CREATE_FAILED';

export const API_WILL_UPDATE = 'AR_API_WILL_UPDATE';
export const API_UPDATE_DONE = 'AR_API_UPDATE_DONE';
export const API_UPDATE_FAILED = 'AR_API_UPDATE_FAILED';

export const API_WILL_DELETE = 'AR_API_WILL_DELETE';
export const API_DELETE_DONE = 'AR_API_DELETE_DONE';
export const API_DELETE_FAILED = 'AR_API_DELETE_FAILED';

export const API_WILL_INDEX = 'AR_API_WILL_INDEX';
export const API_INDEX_DONE = 'AR_API_INDEX_DONE';
export const API_INDEX_CLEAR = 'AR_API_INDEX_CLEAR';
