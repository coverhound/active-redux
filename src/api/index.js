import { createReducer } from './utils';
import baseReducer from './reducer';

export * from './reducer';

/**
 * API Reducer
 * @function
 * @alias module:active-redux/api
 * @example
 * import { createStore, combineReducers, applyMiddleware } from 'redux';
 * import { api } from 'active-redux';
 *
 * const initialState = {};
 *
 * export default createStore(
 *   combineReducers({ api: api.reducer }),
 *   initialState,
 *   applyMiddleware(thunk),
 * );
 *
 * @example
 * import { createStore, combineReducers, applyMiddleware } from 'redux';
 * import { api } from 'active-redux';
 *
 * const initialState = {};
 *
 * // You can add your own extensions to the reducer
 * const customExtensions = [{
 *   map: { AR_PAGINATE_MODELS: (state, payload) => {} },
 *   initialState: { pagination: {} },
 * }];
 *
 * export default createStore(
 *   combineReducers({ api: api.reducer.extend(...customExtension) }),
 *   initialState,
 *   applyMiddleware(thunk),
 * );
 * @param {Object} state
 * @param {Object} action
 */
export const reducer = createReducer(baseReducer);
