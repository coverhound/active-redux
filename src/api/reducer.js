/**
 * @module active-redux/api
 */

import imm from 'object-path-immutable';

import {
  createAction,
  apiClient,
  clearResources,
  decrementProperty,
  incrementProperty,
  markPendingResources,
  mergeResources,
  resourcesArray,
} from './utils';

const initialState = {
  resources: {},
  apiConfig: {},
  isCreating: 0,
  isReading: 0,
  isUpdating: 0,
  isDeleting: 0,
};

/**
 * Action constants
 */
const API_CONFIGURE = 'AR_API_CONFIGURE';
const API_CLEAR = 'AR_API_CLEAR';
const API_HYDRATE = 'API_HYDRATE';

const API_WILL_READ = 'AR_API_WILL_READ';
const API_READ_DONE = 'AR_API_READ_DONE';
const API_READ_FAILED = 'AR_API_READ_FAILED';

const API_WILL_CREATE = 'AR_API_WILL_CREATE';
const API_CREATE_DONE = 'AR_API_CREATE_DONE';
const API_CREATE_FAILED = 'AR_API_CREATE_FAILED';

const API_WILL_UPDATE = 'AR_API_WILL_UPDATE';
const API_UPDATE_DONE = 'AR_API_UPDATE_DONE';
const API_UPDATE_FAILED = 'AR_API_UPDATE_FAILED';

const API_WILL_DELETE = 'AR_API_WILL_DELETE';
const API_DELETE_DONE = 'AR_API_DELETE_DONE';
const API_DELETE_FAILED = 'AR_API_DELETE_FAILED';

/**
 * Actions
 */
const wrappedApiRequest = ({
  resource,
  options,
  endpoint,
  dispatch,
  success,
  failure,
}) => apiClient(endpoint, options).then((json) => {
  dispatch(success(json));
  return Promise.resolve(json);
}).catch((error) => {
  const err = error;
  err.resource = resource;

  dispatch(failure(err));
  return Promise.reject(err);
});

/**
 * Configure the API
 * @see {@link https://github.com/mzabriskie/axios#request-config}
 * @function
 * @param {Object} config - Axios configuration
 */
export const apiConfigure = createAction(API_CONFIGURE);

/**
 * Clear an entity from the store
 * @function
 * @private
 * @example
 * import { api } from 'active-redux';
 *
 * state.api.people
 * // => Object<String: Person>
 *
 * dispatch(api.apiClear('people'))
 *
 * state.api.people
 * // => {}
 * @param {Object} data - JSON-API data
 */
export const apiClear = createAction(API_CLEAR);

/**
 * Hydrate the store from JSON-API
 * @function
 * @example
 * import { api } from 'active-redux';
 *
 * state.api.people
 * // => {}
 *
 * const data = [{
 *   type: 'people',
 *   id: '5',
 *   attributes: {
 *     name: 'Joe',
 *     age: 35,
 *   }
 * }];
 *
 * dispatch(api.apiHydrate(data))
 *
 * state.api.people
 * // => Object<String: Person>
 * @param {Object} data - JSON-API data
 */
export const apiHydrate = createAction(API_HYDRATE);

const apiWillCreate = createAction(API_WILL_CREATE);
const apiCreateDone = createAction(API_CREATE_DONE);
const apiCreateFailed = createAction(API_CREATE_FAILED);

/**
 * Create a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<null>
 *
 * const person = new Person({ attributes: { name: 'Joe' } });
 *
 * dispatch(api.apiCreate({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<Person>
 * @param {Object} args
 * @param {define~Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiCreate = ({ resource, endpoint = resource.endpoint('create') }) => (dispatch, getState) => {
  dispatch(apiWillCreate(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().api.apiConfig,
      method: 'POST',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: apiCreateDone,
    failure: apiCreateFailed,
  });
};

const apiWillRead = createAction(API_WILL_READ);
const apiReadDone = createAction(API_READ_DONE);
const apiReadFailed = createAction(API_READ_FAILED);

/**
 * Read a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.all();
 * // => Promise<[]>
 *
 * dispatch(api.apiRead({ resource: Person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.all();
 * // => Promise<Array<Person>>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiRead = ({ resource, endpoint = resource.endpoint('read') }) => (dispatch, getState) => {
  dispatch(apiWillRead(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().api.apiConfig,
      method: 'GET',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: apiReadDone,
    failure: apiReadFailed,
  });
};

const apiWillUpdate = createAction(API_WILL_UPDATE);
const apiUpdateDone = createAction(API_UPDATE_DONE);
const apiUpdateFailed = createAction(API_UPDATE_FAILED);

/**
 * Update a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5' name='Joe')>
 *
 * const person = new Person({ id: '5', attributes: { name: 'Jimmy' } });
 *
 * dispatch(api.apiUpdate({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5' name='Jimmy')>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiUpdate = ({ resource, endpoint = resource.endpoint('update') }) => (
  (dispatch, getState) => {
    dispatch(apiWillUpdate(resource));
    return wrappedApiRequest({
      resource,
      options: {
        ...getState().api.apiConfig,
        method: 'PATCH',
        data: { data: resource }
      },
      endpoint,
      dispatch,
      success: apiUpdateDone,
      failure: apiUpdateFailed,
    });
  }
);

const apiWillDelete = createAction(API_WILL_DELETE);
const apiDeleteDone = createAction(API_DELETE_DONE);
const apiDeleteFailed = createAction(API_DELETE_FAILED);

/**
 * Delete a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5')>
 *
 * const person = new Person({ id: '5' });
 *
 * dispatch(api.apiDelete({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<null>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiDelete = ({ resource, endpoint = resource.endpoint('delete') }) => (
  (dispatch, getState) => {
    const options = {
      ...getState().api.apiConfig,
      method: 'DELETE',
    };

    dispatch(apiWillDelete(resource));
    return apiClient(endpoint, options).then((json) => {
      dispatch(apiDeleteDone(resource));
      return Promise.resolve(json);
    }).catch((error) => {
      const err = error;
      err.resource = resource;

      dispatch(apiDeleteFailed(err));
      return Promise.reject(err);
    });
  }
);

const map = {
  [API_CONFIGURE]: (state, { payload: apiConfig }) => (
    imm(state).set('apiConfig', apiConfig).value()
  ),
  [API_CLEAR]: (state, { payload: type }) => (
    clearResources(state, type)
  ),
  [API_HYDRATE]: (state, { payload: resources }) => (
    mergeResources(state, resources)
  ),

  [API_WILL_CREATE]: (state) => (
    incrementProperty(state, 'isCreating')
  ),
  [API_CREATE_DONE]: (state, { payload: resources }) => {
    const newState = mergeResources(state, resources);
    return decrementProperty(newState, 'isCreating');
  },
  [API_CREATE_FAILED]: (state) => (
    decrementProperty(state, 'isCreating')
  ),

  [API_WILL_READ]: (state) => (
    incrementProperty(state, 'isReading')
  ),
  [API_READ_DONE]: (state, { payload: resources }) => {
    const newState = mergeResources(state, resources);
    return decrementProperty(newState, 'isReading');
  },
  [API_READ_FAILED]: (state) => (
    decrementProperty(state, 'isReading')
  ),

  [API_WILL_UPDATE]: (state, { payload: resources }) => {
    const newState = incrementProperty(state, 'isUpdating');
    return markPendingResources(newState, resources);
  },
  [API_UPDATE_DONE]: (state, { payload: resources }) => {
    const newState = mergeResources(state, resources);
    return decrementProperty(newState, 'isUpdating');
  },
  [API_UPDATE_FAILED]: (state) => (
    decrementProperty(state, 'isUpdating')
  ),

  [API_WILL_DELETE]: (state, { payload: resources }) => {
    const newState = incrementProperty(state, 'isDeleting');
    return markPendingResources(newState, resources);
  },
  [API_DELETE_DONE]: (state, { payload: resources }) => {
    const newState = clearResources(state, resourcesArray(resources.data));
    return decrementProperty(newState, 'isDeleting');
  },
  [API_DELETE_FAILED]: (state) => (
    decrementProperty(state, 'isDeleting')
  ),
};

export default { initialState, map };
