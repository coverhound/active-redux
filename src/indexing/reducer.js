import imm from 'object-path-immutable';
import { createAction, resourcesArray } from '../api/utils';

/**
 * Helpers
 */
class Index extends Array {
  constructor({ hash, isFetching = false }, ...args) {
    super(...args);
    this.hash = hash;
    this.isFetching = isFetching;
  }
}

const indexArray = (options, data) => new Index(
  options,
  ...resourcesArray(data).map(({ id, type }) => ({ id, type })),
);

/**
 * Action Names
 */
const API_WILL_INDEX = 'AR_API_WILL_INDEX';
const API_INDEX_DONE = 'AR_API_INDEX_DONE';
const API_INDEX_CLEAR = 'AR_API_INDEX_CLEAR';

/**
 * Actions
 */
const apiWillIndex = createAction(API_WILL_INDEX);
const apiIndexDone = createAction(API_INDEX_DONE);
export const apiIndexClear = createAction(API_INDEX_CLEAR);

export const apiIndexAsync = ({ hash, promise }) => (dispatch, getState) => {
  const indexExists = () => getState().api.indices[hash];

  if (indexExists() && indexExists().isFetching) return promise;

  dispatch(apiWillIndex(hash));

  return promise.then((data) => {
    if (indexExists()) dispatch(apiIndexDone({ hash, data }));
    return data;
  }).catch((err) => {
    dispatch(apiIndexDone({ hash, data: [] }));
  });
};
export const apiIndexSync = ({ hash, data }) => (dispatch) => {
  dispatch(apiWillIndex(hash));
  dispatch(apiIndexDone({ hash, data }));
};

const initialState = {
  indices: {},
};

const map = {
  [API_WILL_INDEX]: (state, { payload: hash, data = state.indices[hash] || [] }) => (
    imm.set(state, ['indices', hash], indexArray({ isFetching: true, hash }, data))
  ),
  [API_INDEX_DONE]: (state, { payload: { hash, data } }) => (
    imm.set(state, ['indices', hash], indexArray({ hash }, data))
  ),
  [API_INDEX_CLEAR]: (state, { payload: hash }) => (
    imm.set(state, ['indices', hash], indexArray({ hash }, []))
  ),
};

/**
 * Active-Redux Reducer Extension
 */
export default { initialState, map };
