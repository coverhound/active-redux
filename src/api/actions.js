import * as ActionTypes from './constants';
import { createAction, apiClient } from './utils';
/**
 * @module active-redux/api
 */

const wrappedApiRequest = ({
  resource,
  options,
  endpoint,
  dispatch,
  success,
  failure,
}) => apiClient(endpoint, options).then((json) => {
  dispatch(success(json));
  return Promise.resolve(json);
}).catch((error) => {
  const err = error;
  err.resource = resource;

  dispatch(failure(err));
  return Promise.reject(err);
});

/**
 * Configure the API
 * @see {@link https://github.com/mzabriskie/axios#request-config}
 * @function
 * @param {Object} config - Axios configuration
 */
export const apiConfigure = createAction(ActionTypes.API_CONFIGURE);

/**
 * Clear an entity from the store
 * @function
 * @private
 * @example
 * import { api } from 'active-redux';
 *
 * state.api.people
 * // => Object<String: Person>
 *
 * dispatch(api.apiClear('people'))
 *
 * state.api.people
 * // => {}
 * @param {Object} data - JSON-API data
 */
export const apiClear = createAction(ActionTypes.API_CLEAR);

/**
 * Hydrate the store from JSON-API
 * @function
 * @example
 * import { api } from 'active-redux';
 *
 * state.api.people
 * // => {}
 *
 * const data = [{
 *   type: 'people',
 *   id: '5',
 *   attributes: {
 *     name: 'Joe',
 *     age: 35,
 *   }
 * }];
 *
 * dispatch(api.apiHydrate(data))
 *
 * state.api.people
 * // => Object<String: Person>
 * @param {Object} data - JSON-API data
 */
export const apiHydrate = createAction(ActionTypes.API_HYDRATE);

const apiWillCreate = createAction(ActionTypes.API_WILL_CREATE);
const apiCreateDone = createAction(ActionTypes.API_CREATE_DONE);
const apiCreateFailed = createAction(ActionTypes.API_CREATE_FAILED);

/**
 * Create a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<null>
 *
 * const person = new Person({ attributes: { name: 'Joe' } });
 *
 * dispatch(api.apiCreate({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<Person>
 * @param {Object} args
 * @param {define~Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiCreate = ({ resource, endpoint = resource.endpoint('create') }) => (dispatch, getState) => {
  dispatch(apiWillCreate(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().api.apiConfig,
      method: 'POST',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: apiCreateDone,
    failure: apiCreateFailed,
  });
};

const apiWillRead = createAction(ActionTypes.API_WILL_READ);
const apiReadDone = createAction(ActionTypes.API_READ_DONE);
const apiReadFailed = createAction(ActionTypes.API_READ_FAILED);

/**
 * Read a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.all();
 * // => Promise<[]>
 *
 * dispatch(api.apiRead({ resource: Person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.all();
 * // => Promise<Array<Person>>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiRead = ({ resource, endpoint = resource.endpoint('read') }) => (dispatch, getState) => {
  dispatch(apiWillRead(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().api.apiConfig,
      method: 'GET',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: apiReadDone,
    failure: apiReadFailed,
  });
};

const apiWillUpdate = createAction(ActionTypes.API_WILL_UPDATE);
const apiUpdateDone = createAction(ActionTypes.API_UPDATE_DONE);
const apiUpdateFailed = createAction(ActionTypes.API_UPDATE_FAILED);

/**
 * Update a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5' name='Joe')>
 *
 * const person = new Person({ id: '5', attributes: { name: 'Jimmy' } });
 *
 * dispatch(api.apiUpdate({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5' name='Jimmy')>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiUpdate = ({ resource, endpoint = resource.endpoint('update') }) => (
  (dispatch, getState) => {
    dispatch(apiWillUpdate(resource));
    return wrappedApiRequest({
      resource,
      options: {
        ...getState().api.apiConfig,
        method: 'PATCH',
        data: { data: resource }
      },
      endpoint,
      dispatch,
      success: apiUpdateDone,
      failure: apiUpdateFailed,
    });
  }
);

const apiWillDelete = createAction(ActionTypes.API_WILL_DELETE);
const apiDeleteDone = createAction(ActionTypes.API_DELETE_DONE);
const apiDeleteFailed = createAction(ActionTypes.API_DELETE_FAILED);

/**
 * Delete a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5')>
 *
 * const person = new Person({ id: '5' });
 *
 * dispatch(api.apiDelete({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<null>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiDelete = ({ resource, endpoint = resource.endpoint('delete') }) => (
  (dispatch, getState) => {
    const options = {
      ...getState().api.apiConfig,
      method: 'DELETE',
    };

    dispatch(apiWillDelete(resource));
    return apiClient(endpoint, options).then((json) => {
      dispatch(apiDeleteDone(resource));
      return Promise.resolve(json);
    }).catch((error) => {
      const err = error;
      err.resource = resource;

      dispatch(apiDeleteFailed(err));
      return Promise.reject(err);
    });
  }
);

const apiWillIndex = createAction(ActionTypes.API_WILL_INDEX);
const apiIndexDone = createAction(ActionTypes.API_INDEX_DONE);
export const apiIndexClear = createAction(ActionTypes.API_INDEX_CLEAR);

export const apiIndexAsync = ({ hash, promise }) => (dispatch, getState) => {
  const indexExists = () => getState().api.indices[hash];

  if (indexExists() && indexExists().isFetching) return promise;

  dispatch(apiWillIndex(hash));

  return promise.then((resources) => {
    if (!indexExists()) return resources;
    dispatch(apiIndexDone({ hash, resources }));
    return resources;
  });
};
export const apiIndexSync = ({ hash, resources }) => (dispatch) => {
  dispatch(apiWillIndex(hash));
  dispatch(apiIndexDone({ hash, resources }));
};
