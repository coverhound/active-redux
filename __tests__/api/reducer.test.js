import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import * as subject from 'active-redux/api/reducer';
import createMockStore from 'helpers/create-mock-store';

describe('reducer', () => {
  const person = {
    type: 'people',
    id: 5,
    attributes: {
      hair: 'black',
      age: 25,
    }
  };

  let store;
  const resource = { endpoint: () => 'people', data: person };

  const fullStore = {
    api: {
      apiConfig: {},
      indices: {},
      isCreating: 0,
      isDeleting: 0,
      isReading: 0,
      isUpdating: 0,
      resources: {
        [person.type]: {
          [person.id]: { attributes: { name: 'Johnny' } },
          2: { attributes: { name: 'Yangster' } }
        }
      },
    },
  };

  beforeEach(() => {
    store = createMockStore();
  });

  describe('initialState', () => {
    it('has an empty resources object', () => {
      expect(store.getState().api).toMatchSnapshot();
    });
  });

  describe('apiConfigure()', () => {
    it('updates the apiConfig key', () => {
      store.dispatch(subject.apiConfigure({ baseUrl: 'http://coverhound.com/api' }));
      expect(store.getState().api).toMatchSnapshot();
    });
  });


  describe('apiHydrate()', () => {
    it('populates the state with resources', () => {
      store.dispatch(subject.apiHydrate(resource));
      expect(store.getState().api).toMatchSnapshot();
    });
  });

  describe('apiClear()', () => {
    beforeEach(() => {
      store = createMockStore(fullStore);
    });

    it('clears the resource', () => {
      store.dispatch(subject.apiClear(person.type));
      expect(store.getState().api).toMatchSnapshot();
    });
  });

  describe('HTTP', () => {
    const httpHeaders = { 'content-type': 'application/json' };
    const mockAxios = new MockAdapter(axios);
    const blankPromise = new Promise(() => {});

    describe('apiCreate()', () => {
      it('increments isCreating', () => {
        mockAxios.onPost('/people').replyOnce(() => blankPromise);

        store.dispatch(subject.apiCreate({ resource }));
        expect(store.getState().api).toMatchSnapshot();
      });

      it('populates the state with resources', () => {
        mockAxios.onPost('/people').replyOnce(201, { data: person }, httpHeaders);

        return store.dispatch(subject.apiCreate({ resource })).then(() => {
          expect(store.getState().api).toMatchSnapshot();
        });
      });
    });

    describe('apiRead()', () => {
      it('increments isReading', () => {
        mockAxios.onGet('/people').replyOnce(() => blankPromise);

        store.dispatch(subject.apiRead({ resource }));
        expect(store.getState().api).toMatchSnapshot();
      });

      it('populates the state with resources', () => {
        mockAxios.onGet('/people').replyOnce(200, { data: person }, httpHeaders);

        return store.dispatch(subject.apiRead({ resource })).then(() => {
          expect(store.getState().api).toMatchSnapshot();
        });
      });
    });

    describe('apiUpdate()', () => {
      it('increments isUpdating and marks the resource as pending', () => {
        mockAxios.onPatch('/people').replyOnce(() => blankPromise);

        store.dispatch(subject.apiUpdate({ resource }));
        expect(store.getState().api).toMatchSnapshot();
      });

      it('updates the state with resources', () => {
        mockAxios.onPatch('/people').replyOnce(200, { data: person }, httpHeaders);

        return store.dispatch(subject.apiUpdate({ resource })).then(() => {
          expect(store.getState().api).toMatchSnapshot();
        });
      });
    });

    describe('apiDelete()', () => {
      beforeEach(() => {
        store = createMockStore(fullStore);
      });

      it('increments isDeleting and marks the resource as pending', () => {
        mockAxios.onDelete('/people').replyOnce(() => blankPromise);

        store.dispatch(subject.apiDelete({ resource }));
        expect(store.getState().api).toMatchSnapshot();
      });

      it('removes the resources from the store', () => {
        mockAxios.onDelete('/people').replyOnce(200, { data: person }, httpHeaders);

        return store.dispatch(subject.apiDelete({ resource })).then(() => {
          expect(store.getState().api).toMatchSnapshot();
        });
      });
    });
  });
});
