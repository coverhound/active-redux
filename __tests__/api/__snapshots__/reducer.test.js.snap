// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer HTTP apiCreate() increments isCreating 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 1,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer HTTP apiCreate() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer HTTP apiDelete() increments isDeleting and marks the resource as pending 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 1,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "name": "Johnny",
        },
        "isPending": true,
      },
    },
  },
}
`;

exports[`reducer HTTP apiDelete() removes the resources from the store 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
    },
  },
}
`;

exports[`reducer HTTP apiRead() increments isReading 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 1,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer HTTP apiRead() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer HTTP apiUpdate() increments isUpdating and marks the resource as pending 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 1,
  "resources": Object {},
}
`;

exports[`reducer HTTP apiUpdate() updates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer apiClear() clears the resource 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {},
  },
}
`;

exports[`reducer apiConfigure() updates the apiConfig key 1`] = `
Object {
  "apiConfig": Object {
    "baseUrl": "http://coverhound.com/api",
  },
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer apiHydrate() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer initialState has an empty resources object 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;
